### PR TITLE
v2-alpha-upload feature addded

### PR DESCRIPTION
**Pull request by yashaswi-imagekit**

### CODE REVIEW

File: src/utils/request.ts [hunk 1]
```diff
@@ -59,6 +81,45 @@ export const request = (
...
+export const getJwt = (
+    authenticationEndpoint: string,
+    jwtRequestOptions: JwtRequestOptions
+): Promise<JwtResponse> => {
+    return new Promise((resolve, reject) => {
+        var xhr = new XMLHttpRequest();
+        xhr.timeout = 60000;
+        var urlObj = new URL(authenticationEndpoint);
+        urlObj.searchParams.set("t", Math.random().toString());
+        xhr.open('POST', urlObj.toString());
```
Comment (abhinavr4) [inline src/utils/request.ts:93]: Directly use the `authenticationEndpoint` string here

File: src/utils/request.ts [hunk 2]
```diff
@@ -41,13 +45,31 @@ const addResponseHeadersAndBody = (body: any, xhr: XMLHttpRequest): IKResponse<U
...
+            }, (ex) => {
+                return respond(true, ex, callback);
+            });
+        }, (ex) => {
+            return respond(true, ex, callback);
+        });
+        return
+    }
+
+    generateSignatureToken(options.authenticationEndpoint,).then((signaturObj) => {
```
Comment (abhinavr4) [inline src/utils/request.ts:68]: Remove `,`

File: src/upload/index.ts [hunk 3]
```diff
@@ -37,20 +49,24 @@ export const upload = (
...
+        reqPayload.uploadPayload[key] = JSON.stringify(uploadOptions.extensions);
       } else if (key === "customMetadata" && typeof uploadOptions.customMetadata === "object" &&
         !Array.isArray(uploadOptions.customMetadata) && uploadOptions.customMetadata !== null) {
         formData.append('customMetadata', JSON.stringify(uploadOptions.customMetadata));
-      } else if(uploadOptions[key] !== undefined) {
+        reqPayload.uploadPayload[key] = JSON.stringify(uploadOptions.customMetadata);
+      } else if (uploadOptions[key] !== undefined) {
         formData.append(key, String(uploadOptions[key]));
+        if (key !== 'file')
+          reqPayload.uploadPayload[key] = String(uploadOptions[key]);
```
Comment (abhinavr4) [inline src/upload/index.ts:66]: Why is this needed? Isn't this case covered in the first if condition?

File: src/utils/request.ts [hunk 4]
```diff
@@ -41,13 +45,31 @@ const addResponseHeadersAndBody = (body: any, xhr: XMLHttpRequest): IKResponse<U
 export const request = (
     uploadFileXHR: XMLHttpRequest,
     formData: FormData,
+    jwtRequestOptions: JwtRequestOptions,
```
Comment (abhinavr4) [inline src/utils/request.ts:48]: This parameter is optional

File: src/interfaces/UploadOptions.ts [hunk 5]
```diff
@@ -104,3 +104,9 @@ export interface UploadOptions {
    */
   xhr?: XMLHttpRequest
 }
+
+export interface JwtRequestOptions {
+  publicKey: string
+  expire: number
+  uploadPayload:{[key: string]: Object}
```
Comment (abhinavr4) [inline src/interfaces/UploadOptions.ts:111]: Fix formatting. Type will be `{ [key: string]: Object | string | boolean }`. Initially all values are input in this form, but we convert values to `string` before sending to `signature/v2`

**Review by abhinavr4 — CHANGES_REQUESTED** [review]

File: src/upload/index.ts (hunk 3 above)
Comment (Yashaswi-1998) [reply src/upload/index.ts:66]: when typeOf uploadOptions.file  is "string" then this will be required

File: src/utils/request.ts (hunk 4 above)
Comment (Yashaswi-1998) [reply src/utils/request.ts:48]: function will always be called with this parameter , iska use karna optional hai

File: src/utils/request.ts [hunk 6]
```diff
@@ -100,10 +161,12 @@ export const generateSignatureToken = (
 
 export const uploadFile = (
     uploadFileXHR: XMLHttpRequest,
-    formData: FormData
+    formData: FormData,
+    apiVersion: undefined | string = undefined,
 ): Promise<IKResponse<UploadResponse> | Error> => {
     return new Promise((resolve, reject) => {
-        uploadFileXHR.open('POST', 'https://upload.imagekit.io/api/v1/files/upload');
+        const uploadUrl = apiVersion && apiVersion === 'v2-alpha' ? 'https://upload.imagekit.io/api/v2-alpha/files/upload' : 'https://upload.imagekit.io/api/v1/files/upload'
```
[the remark this answers is not in the document]
Comment (Yashaswi-1998) [reply src/utils/request.ts:168]: dn

File: src/utils/request.ts (hunk 1 above)
Comment (Yashaswi-1998) [reply src/utils/request.ts:93]: dn

File: src/utils/request.ts [hunk 7]
```diff
@@ -59,6 +81,45 @@ export const request = (
...
 
+export const getJwt = (
+    authenticationEndpoint: string,
+    jwtRequestOptions: JwtRequestOptions
+): Promise<JwtResponse> => {
+    return new Promise((resolve, reject) => {
+        var xhr = new XMLHttpRequest();
+        xhr.timeout = 60000;
+        var urlObj = new URL(authenticationEndpoint);
+        urlObj.searchParams.set("t", Math.random().toString());
```
[the remark this answers is not in the document]
Comment (Yashaswi-1998) [reply src/utils/request.ts:92]: dn

File: src/utils/request.ts [hunk 8]
```diff
@@ -59,6 +81,45 @@ export const request = (
     });
 }
 
+export const getJwt = (
+    authenticationEndpoint: string,
+    jwtRequestOptions: JwtRequestOptions
+): Promise<JwtResponse> => {
+    return new Promise((resolve, reject) => {
+        var xhr = new XMLHttpRequest();
```
[the remark this answers is not in the document]
Comment (Yashaswi-1998) [reply src/utils/request.ts:89]: dn

File: src/utils/request.ts (hunk 2 above)
Comment (Yashaswi-1998) [reply src/utils/request.ts:68]: dn

File: src/interfaces/UploadOptions.ts (hunk 5 above)
Comment (Yashaswi-1998) [reply src/interfaces/UploadOptions.ts:111]: dn

File: src/utils/request.ts [hunk 9]
```diff
@@ -100,10 +161,12 @@ export const generateSignatureToken = (
 
 export const uploadFile = (
     uploadFileXHR: XMLHttpRequest,
-    formData: FormData
+    formData: FormData,
+    apiVersion: undefined | string = undefined,
```
[the remark this answers is not in the document]
Comment (abhinavr4) [reply src/utils/request.ts:163]: Resolved

File: src/utils/request.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (abhinavr4) [reply src/utils/request.ts]: Resolved

File: src/utils/request.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (abhinavr4) [reply src/utils/request.ts]: Resolved

File: src/utils/request.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (abhinavr4) [reply src/utils/request.ts]: Resolved

File: src/utils/request.ts (hunk 1 above)
Comment (abhinavr4) [reply src/utils/request.ts]: Resolved

File: src/utils/request.ts (hunk 2 above)
Comment (abhinavr4) [reply src/utils/request.ts]: Resolved

File: src/upload/index.ts (hunk 3 above)
Comment (abhinavr4) [reply src/upload/index.ts:66]: Resolved

File: src/utils/request.ts (hunk 4 above)
Comment (abhinavr4) [reply src/utils/request.ts:48]: Resolved

File: src/upload/index.ts (hunk 3 above)
Comment (abhinavr4) [reply src/upload/index.ts:66]: Resolved

**Review by abhinavr4 — APPROVED** [review]
